### PR TITLE
Reapply "[analyzer] Accept C library functions from the `std` namespace" again

### PR DESCRIPTION
**Pull request by NagyDonat**
This reapplies 80ab8234ac309418637488b97e0a62d8377b2ecf again, after fixing a name collision warning in the unit tests (see the revert commit 13ccaf9b9d4400bb128b35ff4ac733e4afc3ad1c for details).

### CODE REVIEW

**Review by Szelethus — APPROVED** [review]
Lets hope it works fine this time around :)

**Comment by NagyDonat** [comment]
I'll delay reapplying this commit a bit because I realized that it has minor unintended side effects in MallocChecker.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp [hunk 1]
```diff
@@ -446,8 +448,11 @@ class MallocChecker
        std::bind(&MallocChecker::checkRealloc, _1, _2, _3, false)},
       {{{"g_realloc_n"}, 3}, &MallocChecker::checkReallocN},
       {{{"g_try_realloc_n"}, 3}, &MallocChecker::checkReallocN},
-      {{{"getline"}, 3}, &MallocChecker::checkGetdelim},
-      {{{"getdelim"}, 4}, &MallocChecker::checkGetdelim},
+
+      // NOTE: the following CallDescription also matches the C++ standard
+      // library function std::getdelim(); the callback will filter it out.
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:453]:
```suggestion
      // library function std::getline(); the callback will filter it out.
```

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp [hunk 2]
```diff
@@ -1435,9 +1440,17 @@ void MallocChecker::checkGMallocN0(const CallEvent &Call,
   C.addTransition(State);
 }
 
+static bool isFromStdNamespace(const CallEvent &Call) {
+  const Decl *FD = Call.getDecl();
+  assert(FD && "a CallDescription cannot match a call without a Decl");
+  return (FD->isInStdNamespace());
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1446]:
```suggestion
  return FD->isInStdNamespace();
```

**Review by steakhal — APPROVED** [review]

**Comment by steakhal** [comment]
This fixed tickets in our backlog. Thanks for pushing for this.

**Pull request merged**